sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -527,16 +523,26 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
       // Close all writers.
       for (Map.Entry<DestinationT, Writer<DestinationT, OutputT>> entry : writers.entrySet()) {
         Writer<DestinationT, OutputT> writer = entry.getValue();
-        FileResult<DestinationT> result;
         try {
           // Close the writer; if this throws let the error propagate.
-          result = writer.close();
-          c.output(result);
+          writer.close();
         } catch (Exception e) {
           // If anything goes wrong, make sure to delete the temporary file.
           writer.cleanup();
           throw e;
         }
+        int shardNumber =
+            shardNumberAssignment == ShardAssignment.ASSIGN_WHEN_WRITING
+                ? c.element().getKey().getShardNumber()
+                : UNKNOWN_SHARDNUM;
+        if (windowedWrites) {
```

Comment:
BTW, we simplified the FileBasedSink API at the expense of extra complexity in multiple places in WriteFiles. Is this really a simplification?

Reply:
This is an intermediate step to more simplifications that will come after FileIO - in this case, I removed result from Writer in order to simplify removing DestinationT from Writer (and later also removing UserT), to end up with having Writer be a bare-bones consumer for format-specific records that are directly written into it, e.g. for text files it is strings, without any user or destination types involved.

I think this is also desirable from a readability point of view: shard number, destination etc. are bookkeeping information private to WriteFiles, and it's better if they are managed by WriteFiles rather than scattered across two classes.